File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys

input = sys.stdin.readline

# 집합을 합칠 때
# 같은 대표를 공유하는 모든 원소의 대표 번호를 변경해줘야함

def find(x):
    while parent[x]!=x:
        parent[x] = parent[parent[x]] # 6의 대표는 7 -> 7의 대표는 1 -> 6의 대표는 1로 초기화
        x = parent[x] # 가장 최상위 대표로 초기화된 x를 저장후 리턴
    return x

def union(a,b):
    ra,rb = find(a),find(b) # 대표자 번호를 가져옴
    if ra != rb: # 대표자 번호가 서로 다르다면 ex) 0,1,1 의 경우 a,b가 같아서 합쳐줄 필요가 없다.
        parent[rb] = ra # 대표자 번호를 한쪽으로 몰아준다.


n,m = map(int,input().split())

parent = [i for i in range(n+1)]


for _ in range(m):
    order,a,b = map(int,input().split())
    if order == 1:
         print("YES") if find(a)==find(b) else print("NO")
    else: # 집합 합치기
        union(a,b)

File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.stream.IntStream;

public class Main {
    static int n,m;
    static int[] parent;

    public static int find(int x){
        while (parent[x]!=x) {
            parent[x] = parent[parent[x]]; // 부모의 부모 대표번호로 초기화
            x = parent[x];
        }
        return x; // 부모를 리턴
    }

    public static void union(int a,int b){

        int ra = find(a);
        int rb = find(b);

        if (ra!=rb){
            parent[rb] = ra;
        }
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        parent = IntStream.rangeClosed(0, n+1).toArray();

        for (int i=0;i<m;i++){
            st = new StringTokenizer(br.readLine());
            int order = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (order == 1){
                int new_a = find(a);
                int new_b = find(b); 
                if (new_a == new_b){
                    System.out.println("YES");
                } else{
                    System.out.println("NO");
                }
            } else{
                union(a, b);
            }

        }
    }
    
}